spec/controllers/user_journey_controller_spec.rb

```diff
@@ -130,22 +132,99 @@
       expect(response).to have_http_status(:success)
       expect(subject).to render_template(:check_your_certificate)
     end
+
+    it 'renders upload certificate page when cert file uploaded' do
+      certmgr_stub_auth
+      post :submit,
+           params: {
+             component_type: msa_component.component_type,
+             component_id: msa_component.id,
+             certificate_id: msa_encryption_cert.id,
+             certificate: {
+               cert_file: upload_file(
+                 name: 'valid.pem',
+                 type: CertificateExtractor::MIME_PEM,
+                 content: PKI.new.generate_signed_cert.to_pem
+               )
+             }
+           }
+
+      expect(response).to have_http_status(:success)
+      expect(subject).to render_template(:check_your_certificate)
+    end
   end
 
   describe '#confirmation' do
     it 'renders confirmation page' do
       certmgr_stub_auth
       certificate = create(:msa_encryption_certificate)
       msa_component = certificate.component
-      post(:confirm,
+      post :confirm,
            params: {
              component_type: msa_component.component_type,
              component_id: msa_component.id,
              certificate_id: msa_encryption_cert.id,
-             certificate: { new_certificate: certificate.value } }
-          )
+             certificate: { new_certificate: certificate.value }
+           }
       expect(response).to have_http_status(:success)
       expect(subject).to render_template(:confirmation)
     end
   end
+
+  context 'submit with invalid certificate' do
+    it 'should redirect to upload certificate page when certificate is invalid' do
+      certmgr_stub_auth
+      post :submit,
+           params: {
+             component_type: msa_component.component_type,
+             component_id: msa_component.id,
+             certificate_id: msa_encryption_cert.id,
+             certificate: {
+               value: create(
+                 :msa_encryption_certificate,
+                 value: PKI.new.generate_encoded_cert(expires_in: 9000.months)
+               ).value
+             }
+           }
+
+      expect(response).to redirect_to :upload_certificate
+      expect(flash[:error]).to include(I18n.t('certificates.errors.valid_too_long'))
+    end
+
+    it 'should redirect to upload certificate page when certificate file is invalid' do
+      certmgr_stub_auth
+      post :submit,
+           params: {
+             component_type: msa_component.component_type,
+             component_id: msa_component.id,
+             certificate_id: msa_encryption_cert.id,
+             certificate: {
+               cert_file: upload_file(
+                 name: 'invalid.txt',
```

Comment:
we're not testing when the extension is correct (.crt) but the file is not a valid cert... in fact, that fails with an unexpected x509 `header too long` exception

Reply:
the header too long thing causes me sad face. will fix.